Examples-Web/IdleTimeout/src/App.tsx

```diff
@@ -0,0 +1,39 @@
+import './App.scss';
+
+import { useEffect } from 'react';
+
+import CursorManager from 'Components/CursorManager';
+import IdleOverlay from 'Components/IdleOverlay';
+import UseIdleTimeout from 'Hooks/IdleTimeout';
+
+import { ConnectionManager } from './TouchFree/Connection/ConnectionManager';
+import { WebInputController } from './TouchFree/InputControllers/WebInputController';
+
+const App = () => {
+    useEffect(() => {
+        ConnectionManager.init();
+        const controller: WebInputController = new WebInputController();
+
+        new CursorManager();
+
+        return () => {
+            controller.disconnect();
+        };
+    }, []);
+
+    // Must be called after ConnectionManager.init()
+    const isIdle = UseIdleTimeout(10000);
+
+    return (
+        <div className="app">
+            <IdleOverlay state={isIdle ? 'Idle' : 'Inactive'} />
```

Comment:
Slight formatting question - nothing major, just want to get your thoughts. It's quite typical in React components to use a `display` prop as a bool for doing this kind of functionality. As `isIdle` is already a bool, this would mean you could pass the prop straight in line this:

```ts
<IdleOverlay display={isIdle} />
```

I find the terms 'Idle' and 'Inactive' a bit too similar to be opposite functionalities!

Reply:
Generally I think bools are inferior to using string literal type unions from an extensibility perspective since more states can be added. I'm not certain of this opinion yet but it's currently my stance.

I have some questions to ask about this:

- Is there any matching expression or construct that can be used inline? Right now this isn't really extensible because any newly added cases would be assumed equal to 'Inactive'.
- I suspect there's an error that can happen or can be turned on which will catch non-exhaustive matching against the union if it were to be extended. Is this the case and is it commonly enabled or enabled by default? Extending the union would be breaking anyway if so.

Regarding the naming of states, I agree I will change these regardless of answers to above.
Also I will either change both to use a common union type for the possible states or to boolean. I had noticed the inconsistency before too but forgot to fix it.

Reply:
Appears typescript doesn't currently have a nice way of doing matching expressions the way I'm thinking. See here https://dev.to/gvergnaud/bringing-pattern-matching-to-typescript-introducing-ts-pattern-v3-0-o1k

Reply:
As you mentioned above, there isn't a nice way of doing this inline. I also agree that having string literals is nice for extensibility but I am a believer in the "make it as simple as possible" philosophy too.

Using a reducer and a switch statement, we can do something a little more complex and can handle unmatched strings too. ⤵️ Might achieve what you're looking for?

```tsx
// IdleOverlay.tsx
import './IdleOverlay.scss';

import React, { useEffect, useReducer } from 'react';

type StateType = 'Idle' | 'Inactive' | string;

interface IdleOverlayProps {
    state: StateType;
}

const defaultDisplayState = {
    backgroundColour: 'rgba(0,0,0,0)',
    display: 'none',
};

const reducer = (_currentState: typeof defaultDisplayState, state: StateType) => {
    switch(state) {
        case 'Idle':
            return { backgroundColour: 'rgba(0,0,0.95)', display: 'flex' };
        case 'Inactive':
            return { backgroundColour: 'rgba(0,0,0,0)', display: 'none' };
        default:
            throw new Error('Overlay state invalid');
    }
};

const IdleOverlay: React.FC<IdleOverlayProps> = ({ state }) => {
    const [displayState, dispatch] = useReducer(reducer, defaultDisplayState);

    useEffect(() => {
        dispatch(state);
    }, [state]);

    return (
        <div className="overlay" style={{ backgroundColor: displayState.backgroundColour }}>
            <div className="overlay-content" style={{ display: displayState.display }}>
                <h1 style={{ color: 'white' }}>Experience touchless interaction</h1>
            </div>
        </div>
    );
};

export default IdleOverlay;
```

Reply:
How does using a reducer compare to IEFE with a switch? Is there any advantages? Only one I can think of is that the reducer is reusable while an IEFE is inline. Performance? I'd suspect an IEFE is created on every use while this reducer function is likely to be persistent...

Reply:
Bit of performance as you say. Also allows you to pass the dispatch function down as a prop which can be very helpful. I'd have a read of [the reducer hook docs](https://reactjs.org/docs/hooks-reference.html#usereducer) if you haven't already - does a much better job of explaining the benefits of reducers than I can!

Reply:
Didn't like the way the display state object here contains properties for different elements so I decided to modify so that the reducer returns JSX directly. Think it's a bit nicer overall.

Reply:
That is clear as to what will be rendered for the various states. My only thought would be that there's quite a lot of duplication in there - but for two options, that's not really a problem. All looks good to me!